### PR TITLE
QDEC mouse wheel

### PR DESCRIPTION
**Pull request by yuhang-ingchips**
QDEC mouse wheel

### CODE REVIEW

File: examples/hid_mouse/src/main.c [hunk 1]
```diff
@@ -168,6 +169,32 @@ int read_from_flash(void *db, const int max_size)
...
+}
+static void QDEC_Setup(void)
+{
+    QDEC_PclkCfg();
+    SYSCTRL_ClearClkGate(SYSCTRL_ITEM_APB_QDEC);
+    SYSCTRL_ReleaseBlock(SYSCTRL_ITEM_APB_PinCtrl |
+                         SYSCTRL_ITEM_APB_QDEC);
+    PINCTRL_SelQDECIn(21, 22);
+
+    SYSCTRL_SelectQDECClk(SYSCTRL_CLK_HCLK, 100);
```
Comment (zjli-2019) [inline examples/hid_mouse/src/main.c:191]: It does not make sense: after `QDEC_PclkCfg()`, `SYSCTRL_SelectQDECClk(SYSCTRL_CLK_HCLK, 100)`.

File: examples/hid_mouse/src/profile.c [hunk 2]
```diff
@@ -166,6 +168,21 @@ const int8_t delta_xy[][2] =
...
  {7,8},{6,9},{4,9},{4,10},{3,10},{1,11},{1,10}};
 
+#if (INGCHIPS_FAMILY == INGCHIPS_FAMILY_916)
+static uint16_t StepCal(uint16_t preData, uint16_t data, uint8_t *dir)
+{
+    uint16_t step = 0;
+    step = *dir ? (preData - data) : (data - preData);
+    if (data - preData > 60000) {
+        step += 65536;
+    }
```
Comment (zjli-2019) [inline examples/hid_mouse/src/profile.c:178]: Why `> 60000`? 

Why `step += 65536`? `step` is `uint16_t`.

File: examples/hid_mouse/src/profile.c [hunk 3]
```diff
@@ -174,10 +191,24 @@ void mouse_report_movement(void)
...
+        uint8_t dir;
+        data = QDEC_GetData();
+        dir = QDEC_GetDirection();
+        if (data == preData)
+            return;
+        uint16_t step;
+        step = StepCal(preData, data, &dir);
+        preData = data;
+        report.wheel = dir ? step : (0 - (int8_t)step);
+#endif
```
Comment (zjli-2019) [inline examples/hid_mouse/src/profile.c:211]: Use interrupt to trigger wheel report.

File: examples/hid_mouse/src/main.c [hunk 4]
```diff
@@ -172,23 +172,25 @@ int read_from_flash(void *db, const int max_size)
...
 {
+    if ((APB_SYSCTRL->QdecCfg >> 15) & 1)
+        return;
     uint32_t hclk = SYSCTRL_GetHClk();
-    uint32_t qdecClk = SYSCTRL_GetClk(SYSCTRL_ITEM_APB_QDEC);
-    uint8_t div = hclk / qdecClk;
-    if (hclk % qdecClk)
+    uint32_t slowClk = SYSCTRL_GetSlowClk();
+    uint8_t div = hclk / slowClk;
+    if (hclk % slowClk)
```
Comment (zjli-2019) [inline examples/hid_mouse/src/main.c:180]: `SYSCTRL_GetClk(SYSCTRL_ITEM_APB_QDEC)` is still applicative in this case.

File: examples/hid_mouse/src/profile.c [hunk 5]
```diff
@@ -173,9 +173,6 @@ static uint16_t StepCal(uint16_t preData, uint16_t data, uint8_t *dir)
 {
     uint16_t step = 0;
     step = *dir ? (preData - data) : (data - preData);
-    if (data - preData > 60000) {
-        step += 65536;
-    }
     if (step > 60000) {
         *dir ^= 1;
         step = StepCal(preData, data, dir);
```
Comment (zjli-2019) [inline examples/hid_mouse/src/profile.c:178]: Why 60000?

Comment (yuhang-ingchips) [reply examples/hid_mouse/src/profile.c:178]: Is these wrong?

File: examples/hid_mouse/src/main.c (hunk 4 above)
Comment (yuhang-ingchips) [reply examples/hid_mouse/src/main.c:180]: Is these wrong?

**Pull request merged**